hooks/command

```diff
@@ -13,64 +13,152 @@ fi
+  while ! kubectl logs --follow "job/${job_name}" 2>/dev/null;
+  do
+    sleep "$log_loop_retry_interval_sec"
+  done
+
+  echo "0" > "$job_log_complete_marker_file"
 }
 
 echo "--- :kubernetes: Starting Kubernetes Job"
 
 export patchFunc=${BUILDKITE_PLUGIN_K8S_PATCH:-"function(f) f"}
 
-jobspec=$(jsonnet \
+job_spec="$(jsonnet \
   --tla-str "jobName=${job_name}" \
   --tla-str-file "stepEnvFile=${BUILDKITE_ENV_FILE}" \
   --tla-code "agentEnv=$(jq -c -n env)" \
   --tla-code patchFunc \
-  "${basedir}/lib/job.jsonnet")
+  "${basedir}/lib/job.jsonnet")"
 
-export BUILDKITE_PLUGIN_K8S_USE_AGENT_NODE_AFFINITY=${BUILDKITE_PLUGIN_K8S_USE_AGENT_NODE_AFFINITY:-'false'}
-if [[ "$BUILDKITE_PLUGIN_K8S_USE_AGENT_NODE_AFFINITY" == 'true' ]]; then
+if [[ "$use_agent_node_affinity" == "true" ]]; then
   for field in affinity tolerations; do
-    buildkite_agent_value=$(kubectl get pod "$(cat /etc/hostname)" -o json | jq ".spec.$field")
-    jobspec=$(echo "$jobspec" | jq ".spec.template.spec.$field=$buildkite_agent_value")
+    buildkite_agent_value="$(kubectl get pod "$(cat /etc/hostname)" -o json | jq ".spec.$field")"
+    job_spec="$(echo "$job_spec" | jq ".spec.template.spec.$field=$buildkite_agent_value")"
   done
 fi
 
-echo "$jobspec" | kubectl apply -f -
+if [[ "$print_resulting_job_spec" == "true" ]]; then
+  echo -e "Resulting k8s job spec:\n$job_spec"
+fi
+
+readonly job_apply_start_time="$SECONDS"
+job_apply_exit_code=""
+
+while [[ "$((SECONDS - job_apply_start_time))" -lt "$job_apply_timeout_sec" ]]
+do
+  set +e
+  echo "$job_spec" | kubectl apply -f -
+  job_apply_exit_code="$?"
+  set -e
+
+  if [[ "$job_apply_exit_code" == "0" ]]; then
+    break
+  else
+    echo "Attempt to apply the job failed, exit code '$job_apply_exit_code'"
+    sleep "$job_apply_retry_interval_sec"
+  fi
+done
+
+if [[ "$job_apply_exit_code" != "0" ]]; then
+  echo "Failed to apply job, exit code '$job_apply_exit_code'"
+  exit "$job_apply_exit_code"
+fi
 
 echo "Timeout: ${timeout}s"
 
 echo "+++ :kubernetes: Running image: ${BUILDKITE_PLUGIN_K8S_IMAGE}"
 
 tail_logs &
 
-sleeper=2
 counter=${timeout}
 jobstatus=""
 while [[ -z "$jobstatus" ]] ; do
+  set +e
   jobstatus=$(kubectl get job "${job_name}" -o 'jsonpath={.status.conditions[].type}')
-  sleep $sleeper
+  set -e
+
+  if [[ -n "$jobstatus" ]]; then
+    break
+  else
+    sleep "$job_status_retry_interval_sec"
+  fi
+
   if [[ $timeout -gt 0 ]]; then
-    (( counter -= sleeper )) || jobstatus="timeout"
+    (( counter -= job_status_retry_interval_sec )) || jobstatus="timeout"
   fi
 done
 
-echo
 echo "--- :kubernetes: Job status: $jobstatus"
-status=1
+
+# Wait for logs to be fully printed, printing runs in a separate process and we're racing with it.
+readonly log_complete_start_time="$SECONDS"
+while [[ "$(cat "$job_log_complete_marker_file")" != "0" ]] && [[ "$((SECONDS - log_complete_start_time))" -lt "$log_complete_timeout_sec" ]]
+do
+  sleep "$log_complete_retry_interval_sec"
+done
+
+status=""
```

Comment:
```suggestion
status=1
```

we see issues where somehow we end up exiting with this default value:

```
/buildkite/plugins/github-com-artem-zinnatullin-k8s-buildkite-plugin-git-65192202c691578f25d42d04d57fb4b12d5b49c6/hooks/command: line 149: exit: null: numeric argument required
```

so having this more conservative exit code is a bit nicer

Reply:
The `null` is likely coming from setting the `status` to the output of `jq` e.g here https://github.com/EmbarkStudios/k8s-buildkite-plugin/pull/14/files#diff-7a7c160f9b73cfeaf4d8b93aee769fc1R188

@artem-zinnatullin I think we'd want to only set status if `jq` finds a match? one way would be to use enable exit code with `jq -e` so that it exits with non-zero if no match is found and then, only set the status if the exit code is 0.

Reply:
`jq -e` sounds good, annoying that you have to do `set +e; jq -e; set -e;` or write to a file to like `if jq -e > result` to handle both error code and output, but what can you do I guess 😅